src/datanode/src/heartbeat/handler.rs

```diff
@@ -0,0 +1,185 @@
+// limitations under the License.
+
+use std::sync::Arc;
+
+use api::v1::meta::HeartbeatResponse;
+use async_trait::async_trait;
+use catalog::CatalogManagerRef;
+use common_meta::instruction::{Instruction, InstructionReply};
+use table::engine::manager::TableEngineManagerRef;
+use table::requests::{CloseTableRequest, OpenTableRequest};
+use tokio::sync::Mutex;
+
+use crate::error::Result;
+use crate::heartbeat::utils::mailbox_message_to_incoming_message;
+
+mod close_table;
+mod open_table;
+
+pub type IncomingMessage = (MessageMeta, Instruction);
+pub type OutgoingMessage = (MessageMeta, InstructionReply);
+pub struct MessageMeta {
+    pub id: u64,
+    pub subject: String,
+    pub to: String,
+    pub from: String,
+}
+
+#[derive(Default)]
+pub struct HeartbeatMailbox {
+    pub replies: Vec<OutgoingMessage>,
+}
+
+pub type MailboxRef = Arc<Mutex<HeartbeatMailbox>>;
+
+pub type HeartbeatResponseHeadlerExecutorRef = Arc<dyn HeartbeatResponseHeadlerExecutor>;
+
+pub struct HeartbeatResponseHandlerContext {
+    pub mailbox: MailboxRef,
+}
+
+impl From<MailboxRef> for HeartbeatResponseHandlerContext {
+    fn from(mailbox: MailboxRef) -> Self {
+        Self { mailbox }
+    }
+}
+
+#[async_trait]
+pub trait HeartbeatResponseHeadler: Send + Sync {
+    async fn handle(
+        &self,
+        ctx: &mut HeartbeatResponseHandlerContext,
+        resp: &HeartbeatResponse,
+    ) -> Result<()>;
+}
+
+#[async_trait]
+pub trait HeartbeatResponseHeadlerExecutor: Send + Sync {
+    async fn handle(
+        &self,
+        mut ctx: HeartbeatResponseHandlerContext,
+        resp: HeartbeatResponse,
+    ) -> Result<()>;
+}
+
+pub struct NaiveHandlerExecutor {
+    handler: Arc<dyn HeartbeatResponseHeadler>,
+}
+
+impl NaiveHandlerExecutor {
+    pub fn new(handler: Arc<dyn HeartbeatResponseHeadler>) -> Self {
+        Self { handler }
+    }
+}
+
+#[async_trait]
+impl HeartbeatResponseHeadlerExecutor for NaiveHandlerExecutor {
+    async fn handle(
+        &self,
+        mut ctx: HeartbeatResponseHandlerContext,
+        resp: HeartbeatResponse,
+    ) -> Result<()> {
+        self.handler.handle(&mut ctx, &resp).await
+    }
+}
+
+#[derive(Clone)]
+pub struct InstructionHandler {
+    catalog_manager: CatalogManagerRef,
+    table_engine_manager: TableEngineManagerRef,
+}
+
+impl InstructionHandler {
+    pub fn new(
+        catalog_manager: CatalogManagerRef,
+        table_engine_manager: TableEngineManagerRef,
+    ) -> Self {
+        Self {
+            catalog_manager,
+            table_engine_manager,
+        }
+    }
+
+    pub async fn execute(
+        &self,
+        ctx: &mut HeartbeatResponseHandlerContext,
+        (meta, instruction): IncomingMessage,
+    ) -> Result<()> {
+        match instruction {
+            Instruction::OpenTable {
+                catalog,
+                schema,
+                table,
+                table_id,
+                engine,
+                ..
+            } => self.handle_open_table(
```

Comment:
By default, The `handle_open_table` and `handle_close_table`  will spawn a background thread for processing.